### PR TITLE
serialize: add more tests and docstrings

### PR DESCRIPTION
**Pull request by piodul**
This PR aims to improve the quality of the serialize module:

- Adds a couple of tests which check that proper errors are returned in case when serialization fails,
- Fixes a couple of bugs in errors reported from serialization logic - thanks to the aforementioned tests,
- Documents remaining public items in the `types/serialize` module and adds a compile-time warning about missing public items in order to prevent regressions.

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [ ] ~I have adjusted the documentation in `./docs/source/`.~
- [ ] ~I added appropriate `Fixes:` annotations to PR description.~

### CODE REVIEW

File: scylla-cql/src/types/serialize/row.rs [hunk 1]
```diff
@@ -892,6 +901,123 @@ mod tests {
...
+            }
+        ));
+
+        // Serialization of one of the element fails
+        let v = vec!["Ala ma kota", "Kot ma pchły"];
+        let spec = [col("a", ColumnType::Text), col("b", ColumnType::Int)];
+        let err = do_serialize_err(v, &spec);
+        let err = get_ser_err(&err);
+        assert_eq!(err.rust_name, std::any::type_name::<Vec<&str>>());
+        let BuiltinSerializationErrorKind::ColumnSerializationFailed { name, err: _ } = &err.kind;
```
Comment (Lorak-mmk) [inline scylla-cql/src/types/serialize/row.rs:1042]: I didn't know this code is possible - what happends if the variant is different than `ColumnSerializationFailed`? `BuiltinSerializationErrorKind` is `#[non_exhaustive]` so adding a new variant is backwards compatible afaiu

File: scylla-cql/src/types/serialize/mod.rs [hunk 2]
```diff
@@ -7,6 +11,24 @@ pub mod value;
...
+///   one of types with an impl built into the driver fails. It is also returned
+///   from impls generated by the `SerializeRow` macro.
+/// - [`value::BuiltinSerializationError`] is analogous to the above but is
+///   returned from [`SerializeCql::serialize`](value::SerializeCql::serialize)
+///   instead both in the case of builtin impls and impls generated by the
+///   `SerializeCql` macro. It won't be returned by the `Session` directly,
+///   but it might be nested in the [`row::BuiltinSerializationError`].
+/// - [`row::ValueListToSerializeRowAdapterError`] is returned in case when
+///   a list of values encoded with the legacy `ValueList` trait is passed
+///   as an argument to the statement.
```
Comment (Lorak-mmk) [inline scylla-cql/src/types/serialize/mod.rs:31]: Is this always returned or only when serialization fails?

**Review by Lorak-mmk — CHANGES_REQUESTED** [review]

File: scylla-cql/src/types/serialize/row.rs (hunk 1 above)
Comment (piodul) [reply scylla-cql/src/types/serialize/row.rs:1042]: `ColumnSerializationFailed` is the only variant of `BuiltinSerializationErrorKind`, so this pattern is irrefutable and there is no need for `else`. In fact, I tried putting an `else {}` to future-proof the expression, and while `rustc` ate it, `clippy` complained that it is not needed.

As far as `#[non_exhaustive]` goes, it only affects other crates than the current one - which makes sense; if you add an enum variant, you can immediately fix all the uses of the enum in the same crate.

Comment (Lorak-mmk) [reply scylla-cql/src/types/serialize/row.rs:1042]: Wow, I didn't expect Rust to be that smart in this regard, nice.

File: scylla-cql/src/types/serialize/mod.rs (hunk 2 above)
Comment (piodul) [reply scylla-cql/src/types/serialize/mod.rs:31]: Right, the comment doesn't make sense. The error is only returned if "rewriting" the ValueList via SerializeRow interface fails. I'll update.

File: scylla-cql/src/types/serialize/row.rs
[the remark this answers is not in the document]
Comment (piodul) [reply scylla-cql/src/types/serialize/row.rs:1]: I'm not sure where would be the best way to put it and how to formulate it... Let's leave it for a follow up for now.

File: scylla-cql/src/types/serialize/mod.rs (hunk 2 above)
Comment (piodul) [reply scylla-cql/src/types/serialize/mod.rs]: Fixed

**Review by Lorak-mmk — APPROVED** [review]

**Pull request merged**